code/modules/hydroponics/compost_bin.dm

```diff
@@ -0,0 +1,196 @@
+			return
+		if(!user.unEquip(O))
+			return
+
+		O.forceMove(src)
+		/// make biomass from nutriment and plant matter
+		var/plant_biomass = O.reagents.get_reagent_amount("nutriment") + O.reagents.get_reagent_amount("plantmatter")
+		biomass += min(max(plant_biomass, 0.1) * 10,biomass_capacity-biomass)
+		qdel(O)
+		/// start composting after plants are inserted
+		compost()
+		to_chat(user, "<span class='info'>You put [O] in [src].</span>")
+		SStgui.update_uis(src)
+		return TRUE
+	else
+		to_chat(user, "<span class='warning'>You cannot put this in [name]!</span>")
+	SStgui.update_uis(src)
+	update_icon_state()
+
+
+/obj/machinery/compost_bin/ui_data(mob/user)
+	var/list/data = list(
+		"biomass" = biomass,
+		"biomass_capacity" = biomass_capacity,
+		"compost" = compost,
+		"compost_capacity" = compost_capacity,
+		)
+	return data
+
+/obj/machinery/compost_bin/attack_hand(mob/user)
+	ui_interact(user)
+
+
+/obj/machinery/compost_bin/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "CompostBin", "CompostBin", 390, 300, master_ui, state)
+		ui.set_autoupdate(FALSE)
+		ui.open()
+
+
+/// Start composting if there is enough biomass and space for compost
+/obj/machinery/compost_bin/proc/compost()
+	/// Prevents the compost bin from starting to compost again while already composting
+	if(composting == TRUE)
+		return
+	composting = TRUE
+	if(compost >= compost_capacity || biomass <= 0)
+		composting = FALSE
+		return
+	else
+		addtimer(CALLBACK(src, PROC_REF(convert_biomass)), 100)
+
+/// Convert biomass to compost, then continue composting
+/obj/machinery/compost_bin/proc/convert_biomass()
+	//converts 20% of the biomass to compost each cycle, unless there isn't enough comopst space or there is 10 or under biomass
+	var/conversion_amount = max(min(DECAY*biomass,compost_capacity - compost),min(MIN_CONVERSION,biomass))
+	biomass -= conversion_amount
+	compost += conversion_amount
+	update_icon_state()
+	SStgui.update_uis(src)
+	composting = FALSE
+	compost()
+
+/// Checks if there is enough compost to make the desired amount of soil
+/obj/machinery/compost_bin/proc/enough_compost(amount)
+	if(SOIL_COST * amount > compost)
+		return FALSE
+	return TRUE
+
+/// Makes soil from compost
+/obj/machinery/compost_bin/proc/create_soil(amount)
+	/// Creating soil
+	if(!enough_compost(amount))
+		return
+	new SOIL(loc, amount)
+	compost -= SOIL_COST * amount
+	update_icon_state()
+	compost()
+	SStgui.update_uis(src)
+
+/// calls functions according to ui interaction(just making compost for now)
+/obj/machinery/compost_bin/ui_act(action, list/params)
+	if(..())
+		return
+
+	. = TRUE
+	switch(action)
+		if("create")
+			var/amount = clamp(text2num(params["amount"]), 1, 10)
+			create_soil(amount)
```

Comment:
I'm not the best on TGUI, but does this *have* to be a switch even though we only have one action?

Reply:
Not at all, but I wasn't sure whether I'd only have just one. I'm still thinking on leaving it as is just for readability's sake, since it implies this variable could do more, and removing the switch could imply otherwise.